--- a/Makefile.objs
+++ b/Makefile.objs
@@ -62,6 +62,7 @@ common-obj-y += accel.o
 common-obj-y += replay/
 
 common-obj-y += ui/
+common-obj-y += gui-agent/
 common-obj-y += bt-host.o bt-vhci.o
 bt-host.o-cflags := $(BLUEZ_CFLAGS)
 
--- a/vl.c
+++ b/vl.c
@@ -73,6 +73,7 @@ int main(int argc, char **argv)
 #include "net/slirp.h"
 #include "monitor/monitor.h"
 #include "ui/console.h"
+#include "gui-agent/include/qubes-gui-qemu.h"
 #include "ui/input.h"
 #include "sysemu/sysemu.h"
 #include "sysemu/numa.h"
@@ -144,6 +145,7 @@ static int rtc_date_offset = -1; /* -1 m
 QEMUClockType rtc_clock;
 int vga_interface_type = VGA_NONE;
 static int full_screen = 0;
+static int qubesgui_log_level = 0;
 static int no_frame = 0;
 int no_quit = 0;
 static bool grab_on_hover;
@@ -2059,6 +2061,7 @@ typedef enum DisplayType {
     DT_SDL,
     DT_COCOA,
     DT_GTK,
+    DT_QUBESGUI,
     DT_NONE,
 } DisplayType;
 
@@ -2178,6 +2181,24 @@ static DisplayType select_display(const
         error_report("GTK support is disabled");
         exit(1);
 #endif
+    } else if (strstart(p, "qubes-gui", &opts)) {
+        display = DT_QUBESGUI;
+        while (*opts) {
+            const char *nextopt;
+
+            if (strstart(opts, ",log=", &nextopt)) {
+                opts = nextopt;
+                qubesgui_log_level = strtol(opts, &nextopt, 0);
+                if (nextopt == opts) {
+                    goto invalid_qubesgui_args;
+                }
+            } else {
+            invalid_qubesgui_args:
+                error_report("invalid Qubes GUI option string");
+                exit(1);
+            }
+            opts = nextopt;
+        }
     } else if (strstart(p, "none", &opts)) {
         display = DT_NONE;
     } else {
@@ -4533,6 +4554,9 @@ int main(int argc, char **argv, char **e
     case DT_GTK:
         gtk_display_init(ds, full_screen, grab_on_hover);
         break;
+    case DT_QUBESGUI:
+        qubesgui_pv_display_init(qubesgui_log_level);
+        break;
     default:
         break;
     }
